gems/pending/Scvmm/miq_hyperv_disk.rb

```diff
@@ -0,0 +1,176 @@
+    $log.debug "miq_hyperv_disk.seek(#{offset})"
+    case whence
+    when IO::SEEK_CUR
+      @file_offset += offset
+    when IO::SEEK_END
+      @file_offset = @end_byte_addr + offset
+    when IO::SEEK_SET
+      @file_offset = offset
+    end
+    @file_offset
+  end
+
+  def read(size)
+    $log.debug "miq_hyperv_disk.read(#{size})"
+    return nil if @file_offset >= @file_size
+    size = @file_size - @file_offset if (@file_offset + size) > @file_size
+
+    start_sector, start_offset = @file_offset.divmod(@block_size)
+    end_sector                 = (@file_offset + size - 1) / @block_size
+    number_sectors             = end_sector - start_sector + 1
+
+    @file_offset += size
+    bread_cached(start_sector, number_sectors)[start_offset, size]
+  end
+
+  def bread_cached(start_sector, number_sectors)
+    $log.debug "miq_hyperv_disk.bread_cached(#{start_sector}, #{number_sectors})"
+    @block_cache.keys.each do |block_range|
+      next unless block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
+      sector_offset = start_sector - block_range.first
+      buffer_offset = sector_offset * @block_size
+      length        = number_sectors * @block_size
+      if @cache_hits.include?(start_sector)
+        @cache_hits[start_sector] += 1
+      else
+        @cache_hits[start_sector] = 1
+      end
+      return @block_cache[block_range][buffer_offset, length]
+    end
+    block_range               = entry_range(start_sector, number_sectors)
+    @block_cache[block_range] = bread(start_sector, block_range.last - start_sector + 1)
+    if @cache_misses.include?(start_sector)
+      @cache_misses[start_sector] += 1
+    else
+      @cache_misses[start_sector] = 1
+    end
+
+    sector_offset             = start_sector - block_range.first
+    buffer_offset             = sector_offset * @block_size
+    length                    = number_sectors * @block_size
+
+    @block_cache[block_range][buffer_offset, length]
+  end
+
+  def bread(start_sector, number_sectors)
+    $log.debug "miq_hyperv_disk.bread(#{start_sector}, #{number_sectors})"
+    return nil if start_sector > @lba_end
+    number_sectors = @size_in_blocks - start_sector if (start_sector + number_sectors) > @size_in_blocks
+    read_script = <<-READ_EOL
+$file_stream = [System.IO.File]::Open("#{@virtual_disk}", "Open", "Read", "Read")
+$buffer      = New-Object System.Byte[] #{number_sectors * @block_size}
+$file_stream.seek(#{start_sector * @block_size}, 0)
+$file_stream.read($buffer, 0, #{number_sectors * @block_size})
+[System.Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes($buffer))
+$file_stream.Close()
+READ_EOL
+
+    # TODO: Error Handling
+    t1           = Time.now.current
+    encoded_data = @parser.output_to_attribute(@winrm.run_powershell_script(read_script))
+    t2           = Time.now.current
+    buffer       = MemoryBuffer.create(@block_size * number_sectors)
+    buffer       = ""
```

Comment:
Ok, so that is not the intent, but does appear to be what it is doing. The intent is to fill the buffer with the decoded date in the line below.  Let me look at what is really happening.